sklearn/decomposition/tests/test_truncated_svd.py

```diff
@@ -7,98 +7,106 @@
-    assert_array_almost_equal(comp_a[:9], comp_r[:9])
-    assert_array_almost_equal(comp_a[9:], comp_r[9:], decimal=2)
+    assert_allclose(comp_a[:9], comp[:9], rtol=1e-3)
+    assert_allclose(comp_a[9:], comp[9:], rtol=2e-1, atol=1e-2)
 
 
-def test_attributes():
-    for n_components in (10, 25, 41):
-        tsvd = TruncatedSVD(n_components).fit(X)
-        assert_equal(tsvd.n_components, n_components)
-        assert_equal(tsvd.components_.shape, (n_components, n_features))
+@pytest.mark.parametrize("n_components", (10, 25, 41))
+def test_attributes(n_components, X_sparse):
+    n_features = X_sparse.shape[1]
+    tsvd = TruncatedSVD(n_components).fit(X_sparse)
+    assert tsvd.n_components == n_components
+    assert tsvd.components_.shape == (n_components, n_features)
 
 
 @pytest.mark.parametrize('algorithm', ("arpack", "randomized"))
-def test_too_many_components(algorithm):
+def test_too_many_components(algorithm, X_sparse):
+    n_features = X_sparse.shape[1]
     for n_components in (n_features, n_features + 1):
         tsvd = TruncatedSVD(n_components=n_components, algorithm=algorithm)
-        assert_raises(ValueError, tsvd.fit, X)
+        assert_raises(ValueError, tsvd.fit, X_sparse)
 
 
 @pytest.mark.parametrize('fmt', ("array", "csr", "csc", "coo", "lil"))
-def test_sparse_formats(fmt):
-    Xfmt = Xdense if fmt == "dense" else getattr(X, "to" + fmt)()
+def test_sparse_formats(fmt, X_sparse):
+    n_samples = X_sparse.shape[0]
+    Xfmt = (X_sparse.toarray()
+            if fmt == "dense" else getattr(X_sparse, "to" + fmt)())
     tsvd = TruncatedSVD(n_components=11)
     Xtrans = tsvd.fit_transform(Xfmt)
-    assert_equal(Xtrans.shape, (n_samples, 11))
+    assert Xtrans.shape == (n_samples, 11)
     Xtrans = tsvd.transform(Xfmt)
-    assert_equal(Xtrans.shape, (n_samples, 11))
+    assert Xtrans.shape == (n_samples, 11)
 
 
 @pytest.mark.parametrize('algo', ("arpack", "randomized"))
-def test_inverse_transform(algo):
+def test_inverse_transform(algo, X_sparse):
     # We need a lot of components for the reconstruction to be "almost
     # equal" in all positions. XXX Test means or sums instead?
     tsvd = TruncatedSVD(n_components=52, random_state=42, algorithm=algo)
-    Xt = tsvd.fit_transform(X)
+    Xt = tsvd.fit_transform(X_sparse)
     Xinv = tsvd.inverse_transform(Xt)
-    assert_array_almost_equal(Xinv, Xdense, decimal=1)
+    assert_allclose(Xinv, X_sparse.toarray(), rtol=1e-1, atol=2e-1)
 
 
-def test_integers():
-    Xint = X.astype(np.int64)
+def test_integers(X_sparse):
+    n_samples = X_sparse.shape[0]
+    Xint = X_sparse.astype(np.int64)
     tsvd = TruncatedSVD(n_components=6)
     Xtrans = tsvd.fit_transform(Xint)
-    assert_equal(Xtrans.shape, (n_samples, tsvd.n_components))
+    assert Xtrans.shape == (n_samples, tsvd.n_components)
 
 
-def test_explained_variance():
+def test_explained_variance(X_sparse):
     # Test sparse data
     svd_a_10_sp = TruncatedSVD(10, algorithm="arpack")
```

Comment:
Well I mean this is a somewhat large test, because we compare both solutions with 10 vs 20 components, dense vs sparse, different solvers and that explained variance is correct in all case. So there is no easy parametrization here. We can probably split this in 4 tests. I'll look into it.